toolchain/check/import_ref.cpp

```diff
@@ -1811,6 +1798,106 @@ class ImportRefResolver {
          .element_type_id = context_.GetTypeIdForTypeConstant(elem_const_id)});
   }
 
+  // Perform any work that we deferred until the end of the main Resolve loop.
+  auto PerformPendingWork() -> void {
+    while (!pending_generics_.empty() || !pending_specifics_.empty()) {
+      while (!pending_generics_.empty()) {
+        FinishPendingGeneric(pending_generics_.pop_back_val());
+      }
+      while (!pending_specifics_.empty()) {
+        FinishPendingSpecific(pending_specifics_.pop_back_val());
+      }
+    }
+  }
+
+  // Resolves and returns the local contents for an imported instruction block
+  // of constant instructions.
+  auto ResolveLocalInstBlockContents(SemIR::InstBlockId import_block_id)
+      -> llvm::SmallVector<SemIR::InstId> {
+    auto import_block = import_ir_.inst_blocks().Get(import_block_id);
+
+    llvm::SmallVector<SemIR::InstId> inst_ids;
+    inst_ids.reserve(import_block.size());
+    for (auto import_inst_id : import_block) {
+      inst_ids.push_back(
+          context_.constant_values().GetInstId(ResolveOneInst(import_inst_id)));
+    }
+    return inst_ids;
+  }
+
+  // Resolves and returns a local eval block for a region of an imported
+  // generic.
+  auto ResolveLocalEvalBlock(const SemIR::Generic& import_generic,
+                             SemIR::GenericId generic_id,
+                             SemIR::GenericInstIndex::Region region)
+      -> SemIR::InstBlockId {
+    auto import_block_id = import_generic.GetEvalBlock(region);
+    if (!import_block_id.is_valid()) {
+      return SemIR::InstBlockId::Invalid;
+    }
+
+    auto inst_ids = ResolveLocalInstBlockContents(import_block_id);
+    return RebuildGenericEvalBlock(context_, generic_id, region, inst_ids);
+  }
+
+  // Fills in the remaining information in a partially-imported generic.
+  auto FinishPendingGeneric(PendingGeneric pending) -> void {
+    const auto& import_generic = import_ir_.generics().Get(pending.import_id);
+
+    // Note that the generics list can be reallocated by ResolveLocalEvalBlock
+    // importing more generics.
+    auto decl_block_id =
+        ResolveLocalEvalBlock(import_generic, pending.local_id,
+                              SemIR::GenericInstIndex::Region::Declaration);
+    auto* local_generic = &context_.generics().Get(pending.local_id);
+    local_generic->decl_block_id = decl_block_id;
+
+    auto self_specific_id = MakeSelfSpecific(context_, pending.local_id);
+    local_generic->self_specific_id = self_specific_id;
+    pending_specifics_.push_back({.import_id = import_generic.self_specific_id,
+                                  .local_id = self_specific_id});
+
+    auto definition_block_id =
+        ResolveLocalEvalBlock(import_generic, pending.local_id,
+                              SemIR::GenericInstIndex::Region::Definition);
+    local_generic = &context_.generics().Get(pending.local_id);
+    local_generic->definition_block_id = definition_block_id;
```

Comment:
The use of a pointer (and reassignment) looked a little odd to me, maybe:

```suggestion
    auto& local_generic = context_.generics().Get(pending.local_id);
    local_generic.decl_block_id = decl_block_id;

    auto self_specific_id = MakeSelfSpecific(context_, pending.local_id);
    local_generic.self_specific_id = self_specific_id;
    pending_specifics_.push_back({.import_id = import_generic.self_specific_id,
                                  .local_id = self_specific_id});

    // This may invalidate local_generic.
    auto definition_block_id =
        ResolveLocalEvalBlock(import_generic, pending.local_id,
                              SemIR::GenericInstIndex::Region::Definition);
    context_.generics().Get(pending.local_id).definition_block_id = definition_block_id;
```

Reply:
I started with something like this, but it seemed error-prone to me and harder to systematically get right. Changed to a different approach, what do you think?